File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/TaxRuleSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;
import java.util.List;

import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.TaxRule;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;

public class TaxRuleSyncAdapter extends BaseConditionalSyncAdapter<TaxRule, Long> {
    public TaxRuleSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, String eventSlug, PretixApi api, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, eventSlug, api, feedback);
    }

    @Override
    public void updateObject(TaxRule obj, JSONObject jsonobj) throws JSONException {
        obj.setEvent_slug(eventSlug);
        obj.setServer_id(jsonobj.getLong("id"));
        obj.setJson_data(jsonobj.toString());
    }

    @Override
    public Iterator<TaxRule> runBatch(List<Long> ids) {
        return store.select(TaxRule.class)
                .where(TaxRule.EVENT_SLUG.eq(eventSlug))
                .and(TaxRule.SERVER_ID.in(ids))
                .get().iterator();
    }

    @Override
    Iterator<Tuple> getKnownIDsIterator() {
        return store.select(TaxRule.SERVER_ID)
                .where(TaxRule.EVENT_SLUG.eq(eventSlug))
                .get().iterator();
    }

    @Override
    String getResourceName() {
        return "taxrules";
    }

    @Override
    Long getId(JSONObject obj) throws JSONException {
        return obj.getLong("id");
    }

    @Override
    Long getId(TaxRule obj) {
        return obj.getServer_id();
    }

    @Override
    TaxRule newEmptyObject() {
        return new TaxRule();
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/TaxRuleSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sqldelight.TaxRule
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import org.json.JSONObject

class TaxRuleSyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    eventSlug: String,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback?,
) : BaseConditionalSyncAdapter<TaxRule, Long>(
    db = db,
    fileStorage = fileStorage,
    eventSlug = eventSlug,
    api = api,
    syncCycleId = syncCycleId,
    feedback = feedback,
) {

    override fun getResourceName(): String = "taxrules"

    override fun getId(obj: TaxRule): Long = obj.server_id!!

    override fun getId(obj: JSONObject): Long = obj.getLong("id")

    override fun getJSON(obj: TaxRule): JSONObject = JSONObject(obj.json_data!!)

    override fun queryKnownIDs(): MutableSet<Long>? {
        val res = mutableSetOf<Long>()
        db.taxRuleQueries.selectServerIdsByEventSlug(eventSlug).execute { cursor ->
            while (cursor.next().value) {
                val id = cursor.getLong(0) ?: throw RuntimeException("id column not available")
                res.add(id)
            }
            QueryResult.Unit
        }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        db.taxRuleQueries.insert(
            event_slug = eventSlug,
            json_data = jsonobj.toString(),
            server_id = jsonobj.getLong("id"),
        )
    }

    override fun update(obj: TaxRule, jsonobj: JSONObject) {
        db.taxRuleQueries.updateFromJson(
            event_slug = eventSlug,
            json_data = jsonobj.toString(),
            id = obj.id,
        )
    }

    override fun delete(key: Long) {
        db.taxRuleQueries.deleteByServerId(key)
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.taxRuleQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<Long>): List<TaxRule> =
        db.taxRuleQueries.selectByServerIdListAndEventSlug(
            server_id = parameterBatch,
            event_slug = eventSlug,
        ).executeAsList()

}
